pkg/apiserver/auditing/types.go

```diff
@@ -0,0 +1,235 @@
+package auditing
+
+import (
+	"bytes"
+	"github.com/google/uuid"
+	"io/ioutil"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/klog"
+	"kubesphere.io/kubesphere/pkg/apiserver/request"
+	"kubesphere.io/kubesphere/pkg/client/listers/auditing/v1alpha1"
+	"kubesphere.io/kubesphere/pkg/utils/iputil"
+	"net/http"
+	"time"
+)
+
+const (
+	DefaultWebhook       = "kube-auditing-webhook"
+	DefaultCacheCapacity = 10000
+	CacheTimeout         = time.Second
+	SendTimeout          = time.Second * 3
+	ChannelCapacity      = 10
+)
+
+type Auditing interface {
+	Enable() bool
+	K8sAuditingEnable() bool
+	LogRequestObject(req *http.Request) *Event
+	LogResponseObject(e *Event, resp *ResponseCapture, info *request.RequestInfo)
+}
+
+type Event struct {
+	//The workspace which this audit event happened
+	Workspace string
+	//The devops project which this audit event happened
+	Cluster string
+
+	audit.Event
+}
+
+type EventList struct {
+	Items []Event
+}
+
+type auditing struct {
+	lister  v1alpha1.WebhookLister
+	cache   chan *EventList
+	backend *Backend
+}
+
+func NewAuditing(lister v1alpha1.WebhookLister) Auditing {
+
+	a := &auditing{
+		lister: lister,
+		cache:  make(chan *EventList, DefaultCacheCapacity),
+	}
+
+	a.backend = NewBackend(ChannelCapacity, a.cache, SendTimeout)
+	return a
+}
+
+func (a *auditing) getAuditLevel() audit.Level {
+	wh, err := a.lister.Get(DefaultWebhook)
+	if err != nil {
+		klog.Error(err)
```

Comment:
Per our discussion, pls adjust log level here to avoid log errors when audit component is not installed